### PR TITLE
CLI/Import - extending the returned instances by transition_data

### PR DESCRIPTION
**Pull request by rplevka**
every instance of `Import.whatever` will contain `.transition_data` attribute containing a dictionary or a list of dictionaries containing the current Sat5 -> Sat6 entity mappings taken from `${HOME}/.transition_data/whatever*.csv`

The transition_data is a natural part of every executed import and having the data accessible via the object attribute will add much more readability to our unit tests.

### CODE REVIEW

**Comment by rplevka** [comment]
Closes #2685

File: robottelo/cli/import_.py [hunk 1]
```diff
@@ -37,12 +37,41 @@
...
 
+    @staticmethod
+    def read_transition_csv(csv_files):
+        """Process and return remote CSV files.
+
+        <want your AD here? call 00420774238212 :P>
+
+        :param csv_files: <to be updated>
+
+        :returns: <to be updated>
```
Comment (Ichimonji10) [inline robottelo/cli/import_.py]: <_<

File: robottelo/cli/import_.py [hunk 2]
```diff
@@ -89,10 +146,26 @@ def config_file(cls, options=None):
...
+                    ).stdout[:-1]
+                )
+            ]
+            result.transition_data.append(
+                cls.read_transition_csv(
+                    ssh.command(
+                        u'ls -v ${HOME}/.transition_data/puppet_repositories*'
+                    ).stdout[:-1]
+                )
+            )
```
Comment (rplevka) [inline robottelo/cli/import_.py]: Please note, that some subcommands such as this one create and maintain 2+ files.
I return a list of dictionaries for those but this brings some output inconsistency between various subcommands.
I can "encapsulate" the output of the others to the list (making a 1-item list) if you guys think it'll be better.
Thanks for suggestions

File: robottelo/cli/import_.py (hunk 1 above)
Comment (rplevka) [inline robottelo/cli/import_.py]: oh snap :/
brb :D

File: robottelo/cli/import_.py (hunk 2 above)
Comment (Ichimonji10) [inline robottelo/cli/import_.py]: How about a list comprehension?

```python
            result.transition_data = [
                cls.read_transition_csv(ssh.command(command).stdout[:-1])
                for command
                in (
                    u'ls -v ${HOME}/.transition_data/products*'
                    u'ls -v ${HOME}/.transition_data/puppet_repositories*'
                )
            ]
```

File: robottelo/cli/import_.py [hunk 3]
```diff
@@ -37,12 +37,46 @@
...
+        This script should mimic the behavior of Katello's hammer-cli-import
+        - load_persistent_maps() method:
+        https://github.com/Katello/hammer-cli-import/blob/master/lib/
+        hammer_cli_import/persistentmap.rb#L113
+
+        :param csv_files: A list of Strings containing absolute paths of the
+        remote CSV files
+
+        :returns: A Dictionary object holding the key-value pairs of
+        the current state of entity mapping
```
Comment (Ichimonji10) [inline robottelo/cli/import_.py]: I think this is more what you're aiming for:

```
        """Process remote CSV files and transition them to Dictionary

        The result depends on the order of the processed files. This script
        should mimic the behavior of Katello's hammer-cli-import
        `load_persistent_maps()`_ method.

        :param csv_files: A list of Strings containing absolute paths of the
            remote CSV files
        :returns: A Dictionary object holding the key-value pairs of
            the current state of entity mapping

        .. _load_persistent_maps():
            https://github.com/Katello/hammer-cli-import/blob/master/lib/hammer_cli_import/persistentmap.rb#L113

        """
```

File: robottelo/cli/import_.py [hunk 4]
```diff
@@ -110,19 +188,42 @@ def content_view(cls, options=None):
...
+                    ).stdout[:-1]
+                )
+            ]
+            result.transition_data.append(
+                cls.read_transition_csv(
+                    ssh.command(
+                        u'ls -v ${HOME}/.transition_data/repositories*'
+                    ).stdout[:-1]
+                )
+            )
```
Comment (Ichimonji10) [inline robottelo/cli/import_.py]: How about a list comprehension, as shown above?

**Comment by Ichimonji10** [comment]
ACK pending comments.

File: robottelo/cli/import_.py [hunk 5]
```diff
@@ -50,37 +86,65 @@ def activation_key(cls, options=None):
 
         """
         cls.command_sub = 'activation-key'
-        return cls.execute(
+        result = cls.execute(
             cls._construct_command(options),
             output_format='csv'
         )
+        if result.return_code == 0:
+            result.transition_data = cls.read_transition_csv(
```
Comment (elyezer) [inline robottelo/cli/import_.py]: `result` is an instance of SSHCommandResult and by patching it we are breaking the meaning of that object is intended. Also this is breaking what is expected for every CLI helper class and method in the entire automation.

Even though this is very handy I don't agree on creating a special case within the CLI framework. To me makes more sense to create a specialized class that have method for retrieving and returning the trasition_data and that should be called on the test body and not inside a CLI helper.

**Comment by rplevka** [comment]
You're right. How about extending the `Import` class in a way so it will return a list:
[<SSHCommandResult>, transition_data]? I believe that is the correct way of extending the class and at the same time it leaves `SSHCommandResult` untouched.
I don't agree this should be taken care of in the test body:

- the result of the Import command __should__ contain a reference to the created/mapped entity after execution. Otherwise it's just a success/failure message and one needs to perform additional checks in order to find out which entities with which id have been created/mapped
- it is too much redundant code (each unit needs to fetch the `transition_data` - several times)

**Comment by elyezer** [comment]
@rplevka you can add methods with the expected return codes for the CLI stuff and then create other methods like, import_with_trasition_data, do_other_stuff_with_transition_data... those higher level methods will return a tuple with the result and the trasition data. Or even create a Specialized class for this so you can assume that for every method it will return a tuple with the ssh command result and the transition data associated.

What I mean to be on the test body is the explicit call to the method that returns the transition data, not to repeat the entire logic.

**Comment by rplevka** [comment]
- the `basic` methods reverted to the original state
- added supplementary methods `*_with_tr_data`, which return extra `transition_data` inside a tuple with `SSHCommandResult`
- tests edited accordingly
```bash
$ nosetests test_import.py -vs
@test: Check whether all supported Sat5 macros are being ... ok
@test: Import all System Groups from the default data set ... ok
@test: Try to Import Collections with the same name to invoke ... ok
@test: Import all organizations from the default data set ... ok
@test: Import all organizations from the default data set ... ok
@test: Try to Import organizations with the same name to invoke ... ok
@test: Import all 3 users from the default data set (predefined ... ok
@test: Try to Merge users with the same name using 'merge-users' ... ok
@test: Try to Import users with the same name to invoke ... ok
@test: Try to Import all organizations and their users from CSV ... ok
@test: Import all System Groups from the default data set ... ok
@test: Try to Import all organizations from the ... ok
@test: Try to Import all users from the ... ok

----------------------------------------------------------------------
Ran 13 tests in 839.692s

OK
```

**Comment by Ichimonji10** [comment]
The code still looks sane to me. Given that I missed some CLI design issues, though, it would probably be good if someone more familiar with the CLI eyeballed this too.

**Comment by elyezer** [comment]
It looks good. ACK

**Pull request merged**